### PR TITLE
Coarsen and overwritebcs modifications

### PR DESCRIPTION
**Pull request by joanibal**
## Purpose
- Modified coarsen to always keep the end nodes to prevent `coarsen` from creating gaps in the mesh
- Make `overwritebcs` compatible with an array of data

## Type of change

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (non-backwards-compatible fix or feature)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Documentation update
- [ ] Maintenance update
- [ ] Other (please describe)

## Testing
- coarsen a block with an even number of cells. 
- use `overwritebcs` to add an array of data to a bc

## Checklist
_Put an `x` in the boxes that apply._

- [x] I have run `flake8` and `black` to make sure the code adheres to PEP-8 and is consistently formatted
- [x] I have run unit and regression tests which pass locally with my changes
- [x] I have added new tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation

### CODE REVIEW

**Review by marcomangano — APPROVED** [review]
I am not really getting the code changes but if I understand the description correctly this is a nice and long-awaited fix! We don't need to worry about the surface nodes / layers to be necessarily `2n + 1` before coarsening, right?

**Review by marcomangano — COMMENTED** [review]
I am not really getting the code changes but if I understand the description correctly this is a nice and long-awaited fix! We don't need to worry about the surface nodes / layers to be necessarily 2n + 1 before coarsening, right?

**Comment by marcomangano** [comment]
(I accidentally approved, ofc the conflicts must be fixed!)

File: cgnsutilities/cgnsutilities.py
Comment (sseraj) [inline cgnsutilities/cgnsutilities.py:1641]: Shouldn't the `boco` and `b2b` coarsen calls be inside the above if statement?

File: cgnsutilities/cgnsutilities.py
Comment (sseraj) [inline cgnsutilities/cgnsutilities.py:1649]: I don't really understand this part. It seems like the end slicers just need to be lists or tuples of dimension 3, not `slice` objects.

File: cgnsutilities/cgnsutilities.py
Comment (sseraj) [inline cgnsutilities/cgnsutilities.py:2319]: Does this need to change like in the `b2b` function?

File: cgnsutilities/cgnsutilities.py
Comment (sseraj) [inline cgnsutilities/cgnsutilities.py:2327]: So this will just truncate the data, not take every other element?

**Comment by joanibal** [comment]
Yeah, I have some updates that need to be pushed

**Comment by anilyil** [comment]
What is the current status of this PR?

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:1649]: You could probably do it without slicers, but I think this is cleaner. You would need to add if statements for each dimension.  
```
if idx_dim==0:
    fine_mapping[0] = -1
    new_coords[-1, :, :] = coords[fine_mapping]
elif idx_dim==1:    
    fine_mapping[1] = -1
    new_coords[:,-1, :] = coords[fine_mapping]
    
```

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:2319]: yep, fixed in the new version

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:2327]: That is right. The data is stored in a 1-D array, at least for all our meshes. I have added those assumptions as comments.

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:1641]: fixed in the new verison

**Comment by joanibal** [comment]
The PR is ready for review again

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:398]: addressed by new changes

File: cgnsutilities/cgns_utils.py
Comment (anilyil) [inline cgnsutilities/cgns_utils.py:256]: Does it make sense to also say how the nodal data should be ordered here?

File: cgnsutilities/cgnsutilities.py
Comment (anilyil) [reply cgnsutilities/cgnsutilities.py:2455]: I could not see the comment you mentioned btw. If we go with the truncate option we can either add it in the docstring of the coarsen method and maybe even add a warning if we do it.

Would it be too difficult to also take every other element (or even do some sort of restriction operation) instead of truncating it? I would expect for most use cases of the array BCs, you would set them from the python script anyways, but doing it like this looks like its digging a rabbit hole for an incoming student picking up a mesh with an array BC, coarsening it, and then that changing the entire BC value. I get that it may become difficult figuring out the 1d array indexing but it looks like it would be worth the effort if we do it now rather than someone else later.

File: cgnsutilities/cgnsutilities.py
Comment (anilyil) [reply cgnsutilities/cgnsutilities.py:1689]: This looks fine to me; I agree that the slicer approach is a bit cleaner so sure. 

It also looks like the edge cases of corners would also be handled properly, but I am not 100% sure I could imagine the issues that might come up. As long as you thought about the edge corners getting handled properly (e.g. coarsening on both i, j, and k with an even number of cells on all directions), then this is fine.

**Review by anilyil — COMMENTED** [review]
I also reviewed to accelerate this PR a bit. The changes look great in general, just a few comments from me.

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:2455]: Alright,  I updated it to take every other element.

**Comment by anilyil** [comment]
This pr is fine for me as long as @sseraj is also happy. I can approve after he takes a look again because he had a few comments earlier.

File: cgnsutilities/cgnsutilities.py
Comment (eirikurj) [inline cgnsutilities/cgnsutilities.py:1672]: Should we add some type of warning or message that is printed to let the user know if a particular dimension could not be "perfectly" coarsened? The intent seems that when coarsening for instance 1x4 `|-|-|-|` would be coarsened to 1x2 `|-|--|` with a larger second cell. See the attached figure.

**Review by eirikurj — CHANGES_REQUESTED** [review]
Sorry for the late review on this. Please see my comments. I think we should improve the testing a bit more to include all edge cases, and we should also include/update the examples. There seems to be still some issue. The attached figure shows two cubes (the first column) coarsened (the second column). The second case "looses" the last part and becomes "shorter"
cube

File: cgnsutilities/cgnsutilities.py
Comment (joanibal) [reply cgnsutilities/cgnsutilities.py:1672]: A message has been added when a dimension cannot be evenly coarsened.

The shortening issue when coarsening from 4 or 3 to 2 has now been fixed too.

**Comment by joanibal** [comment]
Thanks for finding that bug @eirikurj!
I have fixed it and added a test case to make sure it doesn't break in the future. 
It now coarsens the 3x4x2 block to a 2x2x2 block without shortening it.

File: cgnsutilities/cgns_utils.py
Comment (sseraj) [reply cgnsutilities/cgns_utils.py:255]: I think this is the only unaddressed comment

**Review by sseraj — COMMENTED** [review]
Thanks for these changes. This looks good to me as long as @eirikurj's comments are all addressed

File: cgnsutilities/cgns_utils.py
Comment (joanibal) [reply cgnsutilities/cgns_utils.py]: done

**Comment by joanibal** [comment]
I fixed the typo

**Comment by sseraj** [comment]
Could you also fix the style checks?

**Review by sseraj — DISMISSED** [review]

**Comment by joanibal** [comment]
@nwu63 Do you know why this is being blocked by codecov?

**Comment by sseraj** [comment]
> @nwu63 Do you know why this is being blocked by codecov?

Merging is blocked because @eirikurj requested changes. The codecov check is not required

**Comment by joanibal** [comment]
Ah, thank you

**Comment by sseraj** [comment]
@eirikurj Could you take a look at this again when you get the chance?

**Comment by eirikurj** [comment]
This looks good! One minor thing @joanibal did you plan to add this to the examples?

**Comment by joanibal** [comment]
I put the new mesh file in the examples folder since that is where the other mesh is. 
Should I move it somewhere else?

**Comment by eirikurj** [comment]
@joanibal sorry, the location of the mesh is fine. What I meant is to make an example script that applies the coarsening to that mesh. Same for the overwriting of the BCs (there is one already that could be updated). I know its kind of redundant (and maybe self-explanatory for coarsening), but I think we should try to provide simple examples of all functions and features of the tool, without having to dig through testing code, or relying on extensive examples in doc strings that may end up being inconsistent with what is actually implemented.

**Comment by eirikurj** [comment]
@joanibal I had a second look at `coarsen` and ran into some issues. It seems that the BCs have incorrect PointRange.
See the figure below, using mesh `717_wl_L2.cgns` found in the examples directory.
Let's also make sure to add tests that check that these properties are consistent after coarsening.
cgns_coarsen

**Comment by joanibal** [comment]
oof, that was a big testing coverage miss on my part. 
I have fixed the bug and added a check to the coarsen test for each boco point range.

I have also added an example of how to use the bcoverwrite with an array of bcdata

**Comment by sseraj** [comment]
Can you fix the flake8 checks?

**Review by sseraj — APPROVED** [review]

**Comment by joanibal** [comment]
I'm not sure why the pylint style check is failing.

**Comment by sseraj** [comment]
> I'm not sure why the pylint style check is failing. It doesn't fail when I run it locally.

Are you using the .pylintrc file from our .github repo? In any case, the pylint check is not currently required to merge.

**Comment by marcomangano** [comment]
@anilyil can you approve/merge this? I don't feel 100% confident to check all the changes

**Comment by sseraj** [comment]
@eirikurj has to approve before this can be merged because of the requested changes

**Review by eirikurj — APPROVED** [review]

**Pull request merged**
